tapdb/assets_common.go

```diff
@@ -596,3 +598,91 @@ func maybeUpsertAssetMeta(ctx context.Context, db UpsertAssetStore,
 
 	return assetMetaID, nil
 }
+
+type TapscriptTreeStore interface {
+	UpsertTapscriptTreeRootHash(ctx context.Context,
+		rootHash TapscriptTreeRootHash) (int64, error)
+
+	UpsertTapscriptTreeEdge(ctx context.Context,
+		edge TapscriptTreeEdge) (int64, error)
+
+	UpsertTapscriptTreeNode(ctx context.Context, node []byte) (int64, error)
+
+	FetchTapscriptTreeRoots(ctx context.Context) ([]TapscriptTreeRoot,
+		error)
+
+	FetchTapscriptTree(ctx context.Context,
+		rootHash []byte) ([]TapscriptTreeNode, error)
+
+	DeleteTapscriptTreeEdges(ctx context.Context, rootHash []byte) error
+
+	DeleteTapscriptTreeRoot(ctx context.Context, rootHash []byte) error
+}
+
+func upsertTapscriptTree(ctx context.Context, q TapscriptTreeStore,
+	rootHash []byte, branchOnly bool, nodes [][]byte) error {
+
+	if len(rootHash) != chainhash.HashSize {
+		return fmt.Errorf("root hash must be 32 bytes")
+	}
+
+	nodeCount := len(nodes)
+	switch {
+	case nodeCount == 0:
+		return fmt.Errorf("no tapscript tree nodes provided")
+
+	case branchOnly && nodeCount != 2:
+		return fmt.Errorf("tapscript tree of branches must only have" +
+			"2 nodes")
+
+	// Our node index is an int64, so we can't store a tree with more than
+	// MaxInt64 nodes.
+	case nodeCount > math.MaxInt64:
```

Comment:
I don't think we should allow anything even remotely close to this. Since we probably don't want the user to allow to store a whole BitVM commitment. So maybe something close to `math.MaxInt16` would be more reasonable? They could still just use the branches option if they really need more leaves.

Reply:
I'm fine with `MaxInt16`, and I imagine non-BitVM users aren't using such large trees.